src/persist-txn/src/operator.rs

```diff
@@ -0,0 +1,427 @@
+///   free to downgrade its capability to 8.
+/// - The txns_progress operator learns about the write at 9 (the upper is now
+///   10). It knows that the data shard _WAS_ involved in this, so it forwards
+///   on data from its input until the input has progressed to 10, at which
+///   point it can itself downgrade to 10.
+pub fn txns_progress<K, V, T, D, P, C, G>(
+    passthrough: Stream<G, P>,
+    name: &str,
+    client: impl Future<Output = PersistClient> + 'static,
+    txns_id: ShardId,
+    data_id: ShardId,
+    as_of: T,
+    // TODO(txn): Get rid of these when we have a schemaless WriteHandle.
+    data_key_schema: Arc<K::Schema>,
+    data_val_schema: Arc<V::Schema>,
+) -> (Stream<G, P>, Rc<dyn Any>)
+where
+    K: Debug + Codec,
+    V: Debug + Codec,
+    T: Timestamp + Lattice + TotalOrder + StepForward + Codec64,
+    D: Data + Semigroup + Codec64 + Send + Sync,
+    P: Debug + Data,
+    C: TxnsCodec,
+    G: Scope<Timestamp = T>,
+{
+    let mut builder =
+        AsyncOperatorBuilder::new(format!("txns_progress({})", name), passthrough.scope());
+    let (mut passthrough_output, passthrough_stream) = builder.new_output();
+    let mut passthrough_input =
+        builder.new_input_connection(&passthrough, Pipeline, vec![Antichain::new()]);
+
+    let shutdown_button = builder.build(move |mut capabilities| async move {
+        let mut cap = capabilities.pop().expect("not enough inputs");
+        assert!(capabilities.pop().is_none(), "too many inputs");
+        let client = client.await;
+        let (txns_key_schema, txns_val_schema) = C::schemas();
+        let txns_read = client
+            .open_leased_reader::<C::Key, C::Val, _, _>(
+                txns_id,
+                Arc::new(txns_key_schema),
+                Arc::new(txns_val_schema),
+                Diagnostics::from_purpose("txns_progress"),
+            )
+            .await
+            .expect("schema shouldn't change");
+        let mut txns_cache = TxnsCache::<T, C>::open(txns_read).await;
+
+        txns_cache.update_gt(&as_of).await;
+        let snap = txns_cache
+            .data_snapshot(&data_id, as_of.clone())
+            .expect("data shard is registered");
+        let data_write = client
+            .open_writer::<K, V, T, D>(
+                data_id,
+                data_key_schema,
+                data_val_schema,
+                Diagnostics::from_purpose("data read physical upper"),
+            )
+            .await
+            .expect("schema shouldn't change");
+        let () = snap.unblock_read(data_write).await;
+
+        // We've ensured that the data shard's physical upper is past as_of, so
+        // start by passing through data and frontier updates from the input
+        // until it is past the as_of.
+        let mut read_data_past = as_of;
+        let mut output_progress_exclusive = T::minimum();
+        loop {
+            while let Some(event) = passthrough_input.next_mut().await {
```

Comment:
I _think_ the operator is working correctly, but once `passthrough_input.next_mut()` returns `None`, we know that our inputs have shut down. Might want to do something with that explicitly instead of only relying on the frontier going to `[]`.

Reply:
Ooh yeah, great call. I _think_ I have an idea for what to do here, but it's all pretty subtle, so I left this as a `TODO(txn)` to remind myself to come back to it before turning anything on in prod.